Feature: testing

meme-es opened the pull request:
## Testing
This pull request delivers test specs over the public methods of the project:

### Public methods of Player:
```
#name
#character
#score
```
### Public methods of Game
```
#move
#free?
#board
#ended?
#result
```

me-skit committed:
Add specs to test UserInterface and Player

This commit initializes the files to add specs to test public methods of the
class UserInterface (#play) and Player (#name, #character, #score).

me-skit committed:
Add specs to test public methods of the class Game

This commit adds specs to test public methods of the class Game:
- #move
- #free?
- #board
- #ended?

me-skit committed:
Modify readme info

This commit modifies the readme file adding gameplay info.

maelfosso commented on spec/tic_tac_toe_spec.rb:
```diff
@@ -0,0 +1,132 @@
+  let(:player) { Player.new('Manuel', 'X') }
+
+  describe '#move' do
+    context 'when the method is called' do
+      it 'returns nil always' do
+        expect(game.move(player, 5)).to be(nil)
+      end
+    end
```
Testing the `move` method in your situation will result to compare the board content, not the return value
```ruby
game.move(player, 5)
expect(game.board). to be([....])
```

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello, 
You have written an amazing test, simply to read. Thank for that.
However, there are some changes to do

1. I run the test and I got this error
```ruby
$ rspec spec/
Please introduce the first player's name: F..............

Failures:

  1) UserInterface#play when the method is called returns nil allways
     Failure/Error: new_name = gets.chomp
     
     Errno::EISDIR:
       Is a directory @ io_fillbuf - fd:5 spec/
     # ./lib/userinterface.rb:42:in `gets'
     # ./lib/userinterface.rb:42:in `gets'
     # ./lib/userinterface.rb:42:in `block in choose_name'
     # ./lib/userinterface.rb:40:in `loop'
     # ./lib/userinterface.rb:40:in `choose_name'
     # ./lib/userinterface.rb:33:in `new_player'
     # ./lib/userinterface.rb:6:in `play'
     # ./spec/tic_tac_toe_spec.rb:13:in `block (4 levels) in <top (required)>'

Finished in 0.0074 seconds (files took 0.08168 seconds to load)
15 examples, 1 failure

Failed examples:

rspec ./spec/tic_tac_toe_spec.rb:12 # UserInterface#play when the method is called returns nil allways
```
You can write test for **I/O method**, that is for `UserInterface#play`.

2. Please, move the test of each class in their own file.

3. Read the comment below.

Kindly make those changes and submit a new code review request.
Thanks

Mael FOSSO

me-skit committed:
Move the test of each class in their own file

This commit moves the test of each class in their own file and instead of comparing
the result of the method #board in the class Play now test the type of result. The
test of the method UserInterface#play was removed because there are multiple
I/O with are not inside the method rather calling private methods.

mosesogwo commented on spec/player_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  describe '#name' do
+    context 'when the methodd is called' do
+      it 'returns the string name' do
+        expect(player.name).to be_an_instance_of(String)
```
You should add another test to check if the string returned is 'Manuel'

mosesogwo commented on spec/player_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  describe '#character' do
+    context 'when the methodd is called' do
+      it 'returns the symbol string of the player' do
+        expect(player.character).to be_an_instance_of(String)
```
You should add a test to check that the string returned is exactly the player's character

mosesogwo commented on spec/game_spec.rb:
```diff
@@ -0,0 +1,80 @@
+  describe '#board' do
+    context 'when the method is called' do
+      it 'returns a string with the current board status' do
+        expect(game.board).to be_an_instance_of(String)
```
You should add test to check that the string returned correctly represents the state of the board.

mosesogwo reviewed (changes requested):
**Hi @meme-es, @kbjude :clap:. Good job:+1:, but you haven't effected all the changes requested from the previous review**.
- In testing the **move** method, you should check the status of the board after the move has been made; not just the return value. See the previous reviewer's comment on this.
- In addition to the above, for your player_spec, you should check if the return value is correct and not just the class of the return value. See comments on your spec file below.

_**Kindly make these changes and submit another code review request.**_

**Any question? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

me-skit committed:
Redefine test specs for Player and Game methods

This commit redefines the specs for the Player methods:

- #name: test the returned name of the player
- #character: test the returned symbol of the player
- #score: test the right returned score of the player

For the Game class:

- #move: test the state of the board after a movement
- #board: test the board status at the beginning of the game and after a movement

maelfosso reviewed (approved):
# Congratulations :clap: 

You can merge it :+1: 
Good luck with the next :handshake: 

Mael FOSSO

meme-es commented:
Thank you @maelfosso & @mosesogwo for your help.
